2019-08-time-series/stochastic_volatility.py

```diff
@@ -0,0 +1,140 @@
+
+
+"""
+Multivariate Stochastic Volatility Model with constant correlation
+See Model 2 on page 365 in https://pdfs.semanticscholar.org/fccc/6f4ee933d4330eabf377c08f8b2650e1f244.pdf
+"""
+def model(data):
+    """
+    y = diag(exp(h_t / 2)) * eps_t
+    eps ~ Q L_eps rho_t
+    h_{i+1} = mu + Phi(h_t - mu) + eta
+
+    We do this in log space to convert multiplicative noise to additive noise
+    so we can leverage the GaussianHMM distribution.
+
+    log y_kt = h_kt / 2 + log <L_eps, delta_t>
+            ~= h_kt / 2 + gamma_kt where gamma ~ MVN(0. sigma)
+    and we moment match to compute epsilon.
+
+    :param data: Tensor of the shape ``(batch, timesteps, returns)``
+    :type data: torch.Tensor
+    :returns distribution: posterior distribution over states given data
+    """
+    obs_dim = data.shape[-1]
+    hidden_dim = obs_dim
+    with pyro.plate(len(data)):
+        mu = pyro.param('mu', torch.zeros(hidden_dim))
+        L = pyro.param('L', 0.1 * torch.eye(hidden_dim), constraint=constraints.lower_cholesky)
+        init_dist = dist.MultivariateNormal(mu, scale_tril=L)
+
+        L_eta = pyro.param('L_eta', 0.4 * torch.eye(hidden_dim), constraint=constraints.lower_cholesky)
+        mu_eta = torch.zeros(hidden_dim)
+        trans_matrix = pyro.param('phi', 0.5 * torch.ones(hidden_dim))
+        # this gives us a zero matrix with phi on the diagonal
+        trans_matrix = trans_matrix.diag()
+        trans_dist = dist.MultivariateNormal(mu_eta, scale_tril=L_eta)
+
+        mu_gamma = pyro.param('mu_gamma', torch.zeros(obs_dim))
+        L_gamma = pyro.param('L_gamma', 0.5 * torch.eye(obs_dim), constraint=constraints.lower_cholesky)
+        obs_matrix = torch.eye(hidden_dim, obs_dim)
+        # latent state is h_t - mu
+        obs_dist = dist.MultivariateNormal(-mu_gamma, scale_tril=L_gamma)
+
+        hmm_dist = dist.GaussianHMM(init_dist, trans_matrix, trans_dist, obs_matrix, obs_dist)
+        pyro.sample('obs', hmm_dist, obs=data)
+
+        return hmm_dist.filter(data)
+
+
+def sequential_model(num_samples=10, timesteps=500, hidden_dim=2, obs_dim=2, init_dist=None):
+    """
+    Generate data of shape: (samples, timesteps, obs_dim)
+    where the generative model is defined by:
+        y = exp(h/2) * eps
+        h_{t+1} = mu + Phi (h_t - mu) + eta_t
+    where eps and eta are sampled iid from a MVN distribution
+    """
+    ys = []
+    mu_trans = torch.zeros(hidden_dim)
+    cov_trans = pyro.param('L_eta', 0.2 * torch.eye(hidden_dim, hidden_dim))
+    mu_obs = pyro.param('mu_gamma', torch.zeros(obs_dim))
+    cov_obs = pyro.param('L_gamma', 0.2 * torch.eye(obs_dim, obs_dim))
+    transition = pyro.param('phi', 0.2 * torch.randn(hidden_dim))
+    # this is to generate data as the way model 2 does
+    # we would use the entire transition matrix for model 3
+    transition = transition.diag().expand(num_samples, -1, -1)
+    trans_dist = dist.MultivariateNormal(mu_trans, scale_tril=cov_trans).expand((num_samples,))
```

Comment:
Why don't you use a plate in the calling code? See e.g. bart/forecast.py

Reply:
i thought this would be more clear/explicit which dimensions corresponded to what.  i'll rewrite this model for the tutorial